File: sort/countSort/sort_swiftUI/Sort/Lblbc.swift
// 厦门大学计算机专业 | 前华为工程师
// 专注《零基础学编程系列》  http://lblbc.cn/blog
// 包含：Java | 安卓 | 前端 | Flutter | iOS | 小程序 | 鸿蒙
// 公众号：蓝不蓝编程

import SwiftUI

var array = [2, 1, 5, 4, 3]

struct Lblbc: View {
    @State var arrayStr = ""
    var body: some View {
        VStack {
            Text(arrayStr)
            HStack{
                Button(action: {
                    shuffle()
                    arrayStr = convertToStr()
                })
                {
                    Text("打乱")
                }.padding()
                Button(action: {
                    sort()
                    arrayStr = convertToStr()
                })
                {
                    Text("排序")
                }.padding()
            }
            
        }.onAppear(perform: {
            arrayStr = convertToStr()
        })
        .padding()
    }
}

func shuffle() {
    array = [2, 1, 5, 4, 3];
}

func sort2() {
    for i in 0...array.count-2 {
        for j in 0...array.count-2-i {
            if (array[j] > array[j + 1]) {
                let tmp = array[j + 1];
                array[j + 1] = array[j];
                array[j] = tmp;
            }
        }
    }
}

func sort() {
    var min = array[0]
    var max = min
    for i in 0...array.count-1 {
        if (array[i] > max) {
            max = array[i]
        } else if (array[i] < min) {
            min = array[i]
        }
    }
    
    var counting:[Int] =  [Int](repeating: 0, count: max - min + 1)
    
    for i in 0...array.count-1 {
        counting[array[i] - min] += 1
    }
    var index = -1
    for i in 0...counting.count-1 {
        for _ in 0...counting[i]-1 {
            index += 1
            array[index] = i + min
        }
    }
}

func convertToStr()-> String {
    var result = ""
    for i in 0...array.count-1 {
        result.append(String(array[i])+" ")
    }
    return result
}

struct ContentView_Previews: PreviewProvider {
    static var previews: some View {
        Lblbc()
    }
}

File: sort/countSort/sort_C++/Lblbc.cpp
/// 厦门大学计算机专业 | 前华为工程师
/// 专注《零基础学编程系列》  http://lblbc.cn/blog
/// 包含：Java | 安卓 | 前端 | Flutter | iOS | 小程序 | 鸿蒙
/// 公众号：蓝不蓝编程

#include <iostream>
using namespace std;

void printArray(int *array, int count)
{
	for (int i = 0; i < count; i++)
	{
		cout << array[i] << " ";
	}
}

void sort(int *array, int count)
{
	int min = array[0];
	int max = min;
	for (int i = 1; i < count; i++)
	{
		if (array[i] > max)
		{
			max = array[i];
		}
		else if (array[i] < min)
		{
			min = array[i];
		}
	}
	int countingLen = max - min + 1;
	int *counting = new int[countingLen];
	for (int i = 0; i < count; i++)
	{
		counting[array[i] - min] += 1;
	}
	int index = -1;
	for (int i = 0; i < countingLen; i++)
	{
		for (int j = 0; j < counting[i]; j++)
		{
			index++;
			array[index] = i + min;
		}
	}
}

int main()
{
	int array[5] = {2, 1, 5, 4, 3};
	int count = sizeof(array) / sizeof(int);
	printf("排序前:");
	printArray(array, count);

	sort(array, count);

	printf("排序后:");
	printArray(array, count);
	return 0;
}
